### PR TITLE
#219 ReadOnly flag

### PR DESCRIPTION
**Pull request by GneyHabub**

### CODE REVIEW

**Comment by workshur** [comment]
Please update ticket with description.

**Comment by workshur** [comment]
@GneyHabub LGTM in general but let's do quick research if the react context is more suitable for such features

**Comment by GneyHabub** [comment]
> @GneyHabub LGTM in general but let's do quick research if the react context is more suitable for such features

In my opinion, it is not. The actual information is still stored in the redux and we'll have to retrieve it from there. Using react context will only reduce the number of store accesses by one (schema list and schema details can be merged). This doesn't seem like a worthy improvement to me. 
React docs say that _" Context is primarily used when some data needs to be accessible by **many** components at different nesting levels."_

**Comment by workshur** [comment]
At the same time, context allows this **global** option to be used without complicating containers logic.

> React docs say that " Context is primarily used when some data needs to be accessible by many components at different nesting levels."

From my point of view, our situation fits this description perfectly.

@whotake any thoughts?

**Comment by whotake** [comment]
I agree with @workshur, this is a perfect place for context.

File: kafka-ui-react-app/src/redux/reducers/clusters/selectors.ts [hunk 1]
```diff
@@ -24,3 +24,9 @@ export const getOnlineClusters = createSelector(getClusterList, (clusters) =>
 export const getOfflineClusters = createSelector(getClusterList, (clusters) =>
   clusters.filter(({ status }) => status === ServerStatus.OFFLINE)
 );
+
+export const getClustersReadonlyStatus = (clusterName: string) =>
+  createSelector(
+    getClusterList,
+    (clusters) => clusters.find(({ name }) => name === clusterName)?.readOnly
```
Comment (whotake) [inline kafka-ui-react-app/src/redux/reducers/clusters/selectors.ts:31]: I think it's better to have a more explicit value here. If there is no value, we should return `false`. Any non-boolean value doesn't add any logic, so it's pointless to have `undefined`.

File: kafka-ui-react-app/src/components/contexts/ReadOnlyContext.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+
+const initialValue: { isReadOnly: boolean | undefined } = {
+  isReadOnly: undefined,
+};
+const ReadOnlyContext = React.createContext(initialValue);
```
Comment (whotake) [inline kafka-ui-react-app/src/components/contexts/ReadOnlyContext.ts:6]: It's probably better to call it `ClusterContext` as we store an object there. Another option to use it with only boolean value and call something like this `ClusterReadyOnlyContext`.

File: kafka-ui-react-app/src/components/Topics/Topics.tsx [hunk 3]
```diff
@@ -6,46 +6,51 @@ import EditContainer from 'components/Topics/Edit/EditContainer';
...
 import DetailsContainer from './Details/DetailsContainer';
 import NewContainer from './New/NewContainer';
+import ReadOnlyContext from '../contexts/ReadOnlyContext';
 
 interface Props {
   clusterName: ClusterName;
   isFetched: boolean;
   fetchBrokers: (clusterName: ClusterName) => void;
   fetchTopicsList: (clusterName: ClusterName) => void;
+  isReadOnly?: boolean | undefined;
```
Comment (whotake) [inline kafka-ui-react-app/src/components/Topics/Topics.tsx:16]: Tip: You don't need `undefined ` here. `?` already does that.

File: kafka-ui-react-app/src/components/Topics/List/List.tsx [hunk 4]
```diff
@@ -4,6 +4,7 @@ import Breadcrumb from 'components/common/Breadcrumb/Breadcrumb';
 import { NavLink } from 'react-router-dom';
 import { clusterTopicNewPath } from 'lib/paths';
 import ListItem from './ListItem';
+import _ReadOnlyContext from '../../contexts/ReadOnlyContext';
```
Comment (whotake) [inline kafka-ui-react-app/src/components/Topics/List/List.tsx:7]: Let's use proper naming without `_`. You can use destructuring assignment to extract `isReadOnly` flag without another context variable.

**Review by workshur — APPROVED** [review]

**Review by whotake — APPROVED** [review]

**Pull request merged**